fix(js): Faithfully represent import statements in generic AST

nmote opened the pull request:
Previously, a statement such as `import {x, y} from 'z'` would have been desugared into two separate `ImportFrom` statements. This caused issues with matching, such as #5305, and it also led to the issue described in #6532.

This PR changes the `ImportFrom` node to have a list of imported values instead of a single imported value. That allows us to represent the example above as a single `ImportFrom` node without desugaring. It also allows `import {} from 'z';` to be represented (though it's unclear to me why anybody would ever write that). This change addresses the issues mentioned above.

Fixes #5305

Test plan: Automated tests

PR checklist:

- [x] Purpose of the code is [evident to future readers](https://semgrep.dev/docs/contributing/contributing-code/#explaining-code)
- [x] Tests included or PR comment includes a reproducible test plan
- [x] Documentation is up-to-date
- [x] A changelog entry was [added to changelog.d](https://semgrep.dev/docs/contributing/contributing-code/#adding-a-changelog-entry) for any user-facing change
- [x] Change has no security implications (otherwise, ping security team)

If you're unsure on any of this, please see:

- [Contribution guidelines](https://semgrep.dev/docs/contributing/contributing-code)!
- [One of the more specific guides located here](https://semgrep.dev/docs/contributing/contributing/)

nmote commented:
I just updated this PR with some tests that show some more improved matching behavior. Check out `equivalence_import_variations4.js` and compare to https://semgrep.dev/s/B757

emjin commented on semgrep-core/src/matching/Generic_vs_generic.ml:
```diff
@@ -3003,16 +3003,15 @@ and m_directive a b =
+  | G.ImportFrom (_, _, imports)
+    when List.for_all
+           (function
+             | _, None -> true
```
nit: this would be more readable for me if there were underscored names given to the wildcards. In particular, something like `(_name, Some(_aliases))` would let me know what we're filtering for. It's not the norm in this file though so this is fine

nmote replied:
I've made this change and added a comment. Hopefully that's clearer now.

emjin commented on semgrep-core/src/parsing/pfff/js_to_generic.ml:
```diff
@@ -756,22 +763,22 @@ and require_to_import_in_stmt_opt st =
-                         |> G.d)
-                       |> G.s
+                       (id1, alias_opt)
                    | _ -> raise ComplicatedCase)
```
lmao this exception

emjin reviewed (approved):
I'm not super familiar with Javascript imports, but this seems straightforward and good to me.  I wonder if there are any SR issues this solves, I know they've had friction with our Javascript imports

nmote committed:
fix(js): Faithfully represent import statements in generic AST

Previously, a statement such as `import {x, y} from 'z'` would have been
desugared into two separate `ImportFrom` statements. This caused issues
with matching, such as #5305, and it also led to the issue described in
#6532.

This PR changes the `ImportFrom` node to have a list of imported values
instead of a single imported value. That allows us to represent the
example above as a single `ImportFrom` node without desugaring. It also
allows `import {} from 'z';` to be represented (though it's unclear to
me why anybody would ever write that). This change addresses the issues
mentioned above.

Fixes #5305

Test plan: Automated tests

nmote commented:
Yes, @LewisArdern reported the issue that this closes. Maybe there are others, too.

aryx commented on semgrep-core/src/core/ast/AST_generic.ml:
```diff
@@ -1749,8 +1749,7 @@ and directive_kind =
       * module_name
-      * ident
-      * alias option (* as name alias *)
+      * (ident * alias option (* as name alias *)) list
```
Could we add a comment here explaining why we don't "desugar" in multiple ImportFrom like
we do for variable declarations. I already forgot why we do that ...

nmote replied:
https://github.com/returntocorp/semgrep/pull/6634
